File: go/hash_table.go
package main

import (
	"errors"
	"fmt"
)

type HashNode[K comparable, V any] struct {
	key   K
	value V
	next  *HashNode[K, V]
}

type HashTable[K comparable, V any] struct {
	capacity   int
	size       int
	buckets    []*HashNode[K, V]
	loadFactor float64
}

func NewHashTable[K comparable, V any](initialCapacity int) (*HashTable[K, V], error) {
	if initialCapacity <= 0 {
		return nil, errors.New("initial capacity must be positive")
	}
	return &HashTable[K, V]{
		capacity:   initialCapacity,
		size:       0,
		buckets:    make([]*HashNode[K, V], initialCapacity),
		loadFactor: 0.7,
	}, nil
}

func (ht *HashTable[K, V]) Size() int {
	return ht.size
}

func (ht *HashTable[K, V]) IsEmpty() bool {
	return ht.size == 0
}

func (ht *HashTable[K, V]) Capacity() int {
	return ht.capacity
}

func (ht *HashTable[K, V]) hash(key K) int {
	keyStr := fmt.Sprintf("%v", key)
	hash := 0
	for _, char := range keyStr {
		hash = (hash << 5) - hash + int(char)
	}
	if hash < 0 {
		hash = -hash
	}
	return hash % ht.capacity
}

func (ht *HashTable[K, V]) resize() {
	oldBuckets := ht.buckets
	ht.capacity *= 2
	ht.size = 0
	ht.buckets = make([]*HashNode[K, V], ht.capacity)

	for _, bucket := range oldBuckets {
		current := bucket
		for current != nil {
			ht.Set(current.key, current.value)
			current = current.next
		}
	}
}

func (ht *HashTable[K, V]) Set(key K, value V) {
	if float64(ht.size)/float64(ht.capacity) >= ht.loadFactor {
		ht.resize()
	}

	index := ht.hash(key)
	current := ht.buckets[index]

	for current != nil {
		if current.key == key {
			current.value = value
			return
		}
		current = current.next
	}

	newNode := &HashNode[K, V]{key: key, value: value, next: ht.buckets[index]}
	ht.buckets[index] = newNode
	ht.size++
}

func (ht *HashTable[K, V]) Get(key K) (V, error) {
	var zero V
	index := ht.hash(key)
	current := ht.buckets[index]

	for current != nil {
		if current.key == key {
			return current.value, nil
		}
		current = current.next
	}

	return zero, errors.New("key not found")
}

func (ht *HashTable[K, V]) Has(key K) bool {
	index := ht.hash(key)
	current := ht.buckets[index]

	for current != nil {
		if current.key == key {
			return true
		}
		current = current.next
	}

	return false
}

func (ht *HashTable[K, V]) Delete(key K) bool {
	index := ht.hash(key)
	current := ht.buckets[index]
	var prev *HashNode[K, V] = nil

	for current != nil {
		if current.key == key {
			if prev != nil {
				prev.next = current.next
			} else {
				ht.buckets[index] = current.next
			}
			ht.size--
			return true
		}
		prev = current
		current = current.next
	}

	return false
}

func (ht *HashTable[K, V]) Keys() []K {
	keys := make([]K, 0, ht.size)
	for _, bucket := range ht.buckets {
		current := bucket
		for current != nil {
			keys = append(keys, current.key)
			current = current.next
		}
	}
	return keys
}

func (ht *HashTable[K, V]) Values() []V {
	values := make([]V, 0, ht.size)
	for _, bucket := range ht.buckets {
		current := bucket
		for current != nil {
			values = append(values, current.value)
			current = current.next
		}
	}
	return values
}

func (ht *HashTable[K, V]) Entries() [][2]interface{} {
	entries := make([][2]interface{}, 0, ht.size)
	for _, bucket := range ht.buckets {
		current := bucket
		for current != nil {
			entries = append(entries, [2]interface{}{current.key, current.value})
			current = current.next
		}
	}
	return entries
}

func (ht *HashTable[K, V]) Clear() {
	ht.buckets = make([]*HashNode[K, V], ht.capacity)
	ht.size = 0
}

func (ht *HashTable[K, V]) ForEach(callback func(value V, key K, ht *HashTable[K, V])) {
	for _, bucket := range ht.buckets {
		current := bucket
		for current != nil {
			callback(current.value, current.key, ht)
			current = current.next
		}
	}
}

File: dart/hash_table.dart
class HashNode<K, V> {
  final K key;
  V value;
  HashNode<K, V>? next;

  HashNode(this.key, this.value);
}

class HashTable<K, V> {
  int _capacity;
  int _size = 0;
  List<HashNode<K, V>?> _buckets;
  final double _loadFactor = 0.7;

  HashTable(this._capacity) : _buckets = List.filled(_capacity, null);

  int get size => _size;
  bool get isEmpty => _size == 0;
  int get capacity => _capacity;

  int _hash(K key) {
    String keyStr = key.toString();
    int hash = 0;
    for (int i = 0; i < keyStr.length; i++) {
      hash = (hash << 5) - hash + keyStr.codeUnitAt(i);
    }
    if (hash < 0) hash = -hash;
    return hash % _capacity;
  }

  void _resize() {
    List<HashNode<K, V>?> oldBuckets = _buckets;
    _capacity *= 2;
    _size = 0;
    _buckets = List.filled(_capacity, null);

    for (HashNode<K, V>? bucket in oldBuckets) {
      HashNode<K, V>? current = bucket;
      while (current != null) {
        set(current.key, current.value);
        current = current.next;
      }
    }
  }

  void set(K key, V value) {
    if (_size / _capacity >= _loadFactor) {
      _resize();
    }

    int index = _hash(key);
    HashNode<K, V>? current = _buckets[index];

    while (current != null) {
      if (current.key == key) {
        current.value = value;
        return;
      }
      current = current.next;
    }

    HashNode<K, V> newNode = HashNode(key, value);
    newNode.next = _buckets[index];
    _buckets[index] = newNode;
    _size++;
  }

  V? get(K key) {
    int index = _hash(key);
    HashNode<K, V>? current = _buckets[index];

    while (current != null) {
      if (current.key == key) {
        return current.value;
      }
      current = current.next;
    }

    throw ArgumentError('key not found');
  }

  bool has(K key) {
    int index = _hash(key);
    HashNode<K, V>? current = _buckets[index];

    while (current != null) {
      if (current.key == key) {
        return true;
      }
      current = current.next;
    }

    return false;
  }

  bool delete(K key) {
    int index = _hash(key);
    HashNode<K, V>? current = _buckets[index];
    HashNode<K, V>? prev;

    while (current != null) {
      if (current.key == key) {
        if (prev != null) {
          prev.next = current.next;
        } else {
          _buckets[index] = current.next;
        }
        _size--;
        return true;
      }
      prev = current;
      current = current.next;
    }

    return false;
  }

  List<K> keys() {
    List<K> result = [];
    for (HashNode<K, V>? bucket in _buckets) {
      HashNode<K, V>? current = bucket;
      while (current != null) {
        result.add(current.key);
        current = current.next;
      }
    }
    return result;
  }

  List<V> values() {
    List<V> result = [];
    for (HashNode<K, V>? bucket in _buckets) {
      HashNode<K, V>? current = bucket;
      while (current != null) {
        result.add(current.value);
        current = current.next;
      }
    }
    return result;
  }

  List<List<dynamic>> entries() {
    List<List<dynamic>> result = [];
    for (HashNode<K, V>? bucket in _buckets) {
      HashNode<K, V>? current = bucket;
      while (current != null) {
        result.add([current.key, current.value]);
        current = current.next;
      }
    }
    return result;
  }

  void clear() {
    _buckets = List.filled(_capacity, null);
    _size = 0;
  }

  void forEach(void Function(V value, K key, HashTable<K, V> ht) callback) {
    for (HashNode<K, V>? bucket in _buckets) {
      HashNode<K, V>? current = bucket;
      while (current != null) {
        callback(current.value, current.key, this);
        current = current.next;
      }
    }
  }
}
